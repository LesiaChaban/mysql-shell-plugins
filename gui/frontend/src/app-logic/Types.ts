/*
 * Copyright (c) 2021, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

// TODO: check if we can switch to camel case for those DB data types that use snake case.

/* eslint-disable @typescript-eslint/naming-convention */

import * as codicon from "../components/ui/Codicon";

// Types used in different places in the application (modules, parser, scripting etc.).

export interface IDictionary {
    [key: string]: unknown;
}

// Types of columns in a table column definition, which we support in the application.
// Also used for column affinities in SQLite (instead of column types).
export enum DBDataType {
    Unknown,

    // Integer types.
    TinyInt,
    SmallInt,
    MediumInt,
    Int,                // Any integer like number type (numeric is essentially the same) in SQLite.
    Bigint,
    UInteger,           // Mapped type from the shell.

    // Float types.
    Float,
    Real,               // Any floating point type in SQLite.
    Double,
    Decimal,

    // Binary types.
    Binary,
    Varbinary,

    // Text types.
    Char,
    Nchar,
    Varchar,
    Nvarchar,

    String,             // Mapped type from the shell.
    TinyText,
    Text,               // Any text in SQLite.
    MediumText,
    LongText,

    // Blob types.
    TinyBlob,
    Blob,               // NULL or no (explicit) type in SQLite.
    MediumBlob,
    LongBlob,

    // Date and time types.
    DateTime,
    DateTime_f,
    Date,
    Time,
    Time_f,
    Year,
    Timestamp,
    Timestamp_f,

    // Geometry types.
    Geometry,
    Point,
    LineString,
    Polygon,
    GeometryCollection,
    MultiPoint,
    MultiLineString,
    MultiPolygon,

    // Special types.
    Numeric,            // SQLite only, similar to INTEGER.
    Json,
    Bit,
    Boolean,
    Enum,
    Set,
}

// Describes the number of parameters for a DB data type. It doesn't say anything about the parameter types, though.
export enum ParameterFormatType {
    None = 0,
    One = 1,            // (n)
    OneOrZero = 2,      // [(n)]
    Two = 3,            // (m,n)
    TwoOrOne = 4,       // (m[,n])
    TwoOrZero = 5,      // [(m,n)]
    TwoOrOneOrZero = 6, // [(m[, n])]
    List = 10,          // (m, n, o, p ...)
}

export interface IDBCharacterSet {
    collations: string[];
    defaultCollation: string;
    description: string;
}

// Describes a data type in a database.
export interface IDBDataTypeDetails {
    type: DBDataType;

    characterMaximumLength?: number;
    characterOctetLength?: number;
    dateTimePrecision?: number;
    flags?: string[];
    numericPrecision?: number;
    numericPrecisionRadix?: number;
    numericScale?: number;
    needsQuotes?: boolean;
    parameterFormatType?: ParameterFormatType;
    synonyms?: string[];
}

export interface IColumnInfo {
    name: string;

    // Note: SQLite has no column data types like other RDBMS-es. Instead we store column affinities here.
    dataType: IDBDataTypeDetails;
}

export enum MessageType {
    Error,
    Warning,
    Info,
    Interactive,
    Response,
}

export interface IExecutionInfo {
    type?: MessageType;
    text: string;
}

// Types for requesting a specific dialog and sending back the entered values.

export enum DialogType {
    Prompt = "prompt",        // A simple prompt value dialog, requesting a single value from the user.
    MrsService = "mrsService",// A dialog to create or edit an MRS service.
    MrsSchema = "mrsSchema",
}

export interface IDialogRequest extends IDictionary {
    type: DialogType;
    id?: string;    // An optional id to identify the invocation.
    title?: string; // Optionally used to set a customized dialog title (where supported).

    // Values to configure how the dialog looks like (available options in drop downs etc.).
    parameters?: IDictionary;

    // Values to pre-fill certain elements or additional data for captions, payload etc.
    values?: IDictionary;

    // Additional data for specific use cases.
    data?: IDictionary;
}

export interface IDialogResponse extends IDictionary {
    type: DialogType;
    accepted: boolean;

    values?: IDictionary;

    data?: IDictionary;
}

export interface IStatusbarInfo {
    id: string;
    text?: string;
    icon?: string | codicon.Codicon;
    choices?: Array<{ label: string; data: IDictionary }>;
    visible?: boolean;
    standout?: boolean;
}

/**
 * Defines the structure for password requests sent by various access methods. Passwords are stored by the backend
 * using native OS methods.
 */
export interface IServicePasswordRequest {
    requestId: string;    // The id of the request that asked for the password.
    caption?: string;     // The password dialog title.
    description?: string; // A normal description of what is requested.
    service?: string;     // A human readable string describing the service.
    user?: string;        // The user name for which to get the password.

    payload?: unknown;    // Any other data the caller wants to pass over.
}
