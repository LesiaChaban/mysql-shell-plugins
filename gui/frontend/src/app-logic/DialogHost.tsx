/*
 * Copyright (c) 2021, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import React from "react";

import { IDialogSection, IDialogValues, PasswordDialog, ValueDialogBase, ValueEditDialog } from "../components/Dialogs";
import { Component } from "../components/ui";
import { MrsSchemaDialog } from "../modules/mrs/dialogs/MrsSchemaDialog";

import { MrsServiceDialog } from "../modules/mrs/dialogs/MrsServiceDialog";
import { requisitions } from "../supplement/Requisitions";
import { DialogType, IDialogRequest, IDialogResponse, IDictionary } from "./Types";

// A component to host all application wide accessible dialogs.
export class DialogHost extends Component {

    private runningDialogs = new Set<DialogType>();
    private dialogRefs = new Map<DialogType, React.RefObject<ValueDialogBase>>();

    // The value edit dialog is special, as it uses a different approach for editing complex values.
    private promptDialogRef = React.createRef<ValueEditDialog>();

    public constructor(props: {}) {
        super(props);

        requisitions.register("showDialog", this.showDialog);
    }

    public render(): React.ReactNode {
        const dialogs: React.ReactNode[] = [
            // The password dialog has it's own command handling. Just host it here.
            <PasswordDialog key="passwordDialog" />,

            // The value edit dialog has a different value handling, so it's not added to the dialogRefs list.
            <ValueEditDialog
                key="valueEditDialog"
                caption="Feedback Requested"
                ref={this.promptDialogRef}
                onClose={this.handlePromptDialogClose}
            />,
        ];

        let ref = React.createRef<ValueDialogBase>();
        this.dialogRefs.set(DialogType.MrsService, ref);
        dialogs.push(<MrsServiceDialog
            key="mrsServiceDialog"
            ref={ref as React.RefObject<MrsServiceDialog>}
            onClose={this.handleDialogClose.bind(this, DialogType.MrsService)}
        />);

        ref = React.createRef<ValueDialogBase>();
        this.dialogRefs.set(DialogType.MrsSchema, ref);
        dialogs.push(<MrsSchemaDialog
            ref={ref as React.RefObject<MrsSchemaDialog>}
            onClose={this.handleDialogClose.bind(this, DialogType.MrsSchema)}
        />);

        return (
            <>
                {dialogs}
            </>
        );
    }

    private showDialog = (request: IDialogRequest): Promise<boolean> => {
        // Check if a dialog of the given type is already active.
        // Only one of each type can be active at any time.
        if (!this.runningDialogs.has(request.type)) {
            this.runningDialogs.add(request.type);
            if (request.type === DialogType.Prompt) {
                this.runPromptDialog(request);

                return Promise.resolve(true);
            } else {
                const ref = this.dialogRefs.get(request.type);
                if (ref && ref.current) {
                    ref.current.show(request, request.title);

                    return Promise.resolve(true);
                }
            }
        }

        return Promise.resolve(false);
    };

    /**
     * Configures a value edit dialog with a single section to let the user input a single value.
     * Support entries in the request are:
     * - values.prompt A caption for the input field.
     * - values.payload: A value that is forwarded to the response handler.
     *
     * @param request The request with the data for the dialog.
     */
    private runPromptDialog = (request: IDialogRequest): void => {
        const promptSection: IDialogSection = {
            values: {
                input: {
                    caption: request.values?.prompt as string,
                    value: "",
                    span: 8,
                },
            },
        };

        this.promptDialogRef.current?.show(
            {
                id: request.id,
                sections: new Map<string, IDialogSection>([
                    ["prompt", promptSection],
                ]),
            },
            [],
            { backgroundOpacity: 0.1 },
            "",
            undefined,
            request.data,
        );

    };

    private handleDialogClose = (type: DialogType, accepted: boolean, values?: IDictionary): void => {
        if (this.runningDialogs.has(type)) {
            this.runningDialogs.delete(type);

            const response: IDialogResponse = {
                type,
                accepted,
                values,
            };

            void requisitions.execute("dialogResponse", response);
        }
    };

    private handlePromptDialogClose = (accepted: boolean, values: IDialogValues, data?: IDictionary): void => {
        const type = DialogType.Prompt;
        if (this.runningDialogs.has(type)) {

            this.runningDialogs.delete(type);

            const promptSection = values.sections.get("prompt");
            if (promptSection) {
                const response: IDialogResponse = {
                    type,
                    accepted,
                    values: {
                        input: promptSection.values.input.value as string,
                    },
                    data,
                };

                void requisitions.execute("dialogResponse", response);
            }
        }
    };

}
