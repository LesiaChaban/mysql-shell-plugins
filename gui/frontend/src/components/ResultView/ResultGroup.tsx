/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import React from "react";

//import constructionIcon from "../../assets/images/under-construction.svg";

import { Component, IComponentProperties, IComponentState, ITabviewPage, TabPosition, Tabview } from "../ui";
import { ResultView } from ".";
import { IResultSet, IResultSetRows } from "../../script-execution";

export interface IResultGroupProperties extends IComponentProperties {
    resultSet: IResultSet;

    onResultPageChange?: (requestId: string, currentPage: number, sql: string) => void;
}

interface IResultGroupState extends IComponentState {
    selectedTab: string;
}

export class ResultGroup extends Component<IResultGroupProperties, IResultGroupState> {

    private resultRef = React.createRef<ResultView>();

    public constructor(props: IResultGroupProperties) {
        super(props);

        this.state = {
            selectedTab: "",
        };
    }

    public render(): React.ReactNode {
        const { resultSet, onResultPageChange } = this.props;
        const { selectedTab } = this.state;

        const pages: ITabviewPage[] = [];

        const resultView = <ResultView
            key="resultSet"
            ref={this.resultRef}
            resultSet={resultSet}
            onResultPageChange={onResultPageChange}
        />;

        pages.push({ id: "resultSet", caption: "RS", tooltip: "Result Set", content: resultView });

        // TODO: Temporarily disabled. Add a mode to the tabview where no tab is shown and tabs are switched
        // programmatically.
        /*pages.push({ id: "textResult", caption: "TX", tooltip: "Textual Output", content: icon });
        pages.push({ id: "formEditor", caption: "FE", tooltip: "Form Editor", content: icon });
        pages.push({ id: "fieldTypes", caption: "FT", tooltip: "Field Types", content: icon });
        pages.push({ id: "queryStats", caption: "QS", tooltip: "Query Statistics", content: icon });
        pages.push({ id: "executionPlan", caption: "EP", tooltip: "Execution Plan", content: icon });*/

        return (
            <Tabview
                className="resultGroup"
                stretchTabs={false}
                selectedId={selectedTab !== "" ? selectedTab : "resultSet"}
                tabPosition={TabPosition.Right}
                pages={pages}
                hideSingleTab

                onSelectTab={this.handleTabSelection}
            />
        );
    }

    public async addData(newData: IResultSetRows, replace: boolean): Promise<void> {
        if (this.resultRef.current) {
            await this.resultRef.current.addData(newData, replace);
        }
    }

    private handleTabSelection = (id: string): void => {
        this.setState({ selectedTab: id });
    };
}
