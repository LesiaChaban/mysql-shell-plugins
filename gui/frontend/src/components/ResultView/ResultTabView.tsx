/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import React from "react";

import { Component, IComponentProperties, IComponentState, ITabviewPage, TabPosition, Tabview } from "../ui";
import { IResultSet, IResultSetRows, IResultSets } from "../../script-execution";
import { ResultGroup } from ".";
import { IColumnInfo } from "../../app-logic/Types";

export interface IResultTabViewProperties extends IComponentProperties {
    resultSets: IResultSets;

    onResultPageChange?: (requestId: string, currentPage: number, sql: string) => void;
}

interface IResultTabViewState extends IComponentState {
    selectedTab: string;

    // React refs to the used ResultGroup instances, keyed by the request ID for the result set.
    groupRefs: Map<string, React.RefObject<ResultGroup>>;

    // Result IDs that are marked for purge.
    purgePending: Set<string>;
}

// Holds a collection of result views and other output in a tabbed interface.
export class ResultTabView extends Component<IResultTabViewProperties, IResultTabViewState> {

    public constructor(props: IResultTabViewProperties) {
        super(props);

        this.state = {
            selectedTab: "",
            groupRefs: new Map(),
            purgePending: new Set(),
        };
    }

    public render(): React.ReactNode {
        const { resultSets, onResultPageChange } = this.props;
        const { selectedTab, groupRefs } = this.state;

        const pages = resultSets.sets.map((resultSet: IResultSet, index: number): ITabviewPage => {
            const ref = React.createRef<ResultGroup>();
            groupRefs.set(resultSet.head.requestId, ref);

            return {
                id: `resultGroup${index}`,
                caption: `Result ${index + 1}`,
                content: (
                    <ResultGroup
                        ref={ref}
                        resultSet={resultSet}

                        onResultPageChange={onResultPageChange}
                    />
                ),
            };
        });

        return (
            <Tabview
                className="resultHost"
                stretchTabs={false}
                hideSingleTab={true}
                selectedId={selectedTab !== "" ? selectedTab : "resultGroup0"}
                tabPosition={TabPosition.Top}
                pages={pages}

                onSelectTab={this.handleTabSelection}
            />
        );
    }

    /**
     * Triggers a column update in the group which belongs to the given request id.
     *
     * @param requestId The actual request ID which identifies a specific result group.
     * @param columns The columns to set.
     */
    public async updateColumns(requestId: string, columns: IColumnInfo[]): Promise<void> {
        const { groupRefs } = this.state;

        const groupRef = groupRefs.get(requestId);
        if (groupRef && groupRef.current) {
            await groupRef.current.updateColumns(columns);
        }
    }

    /**
     * In order to update large result sets without re-rendering everything we use direct methods
     * to add new data.
     *
     * @param newData The data to add.
     *
     * @returns A promise the resolves when the operation is complete.
     */
    public async addData(newData: IResultSetRows): Promise<void> {
        const { groupRefs, purgePending } = this.state;

        const needPurge = purgePending.delete(newData.requestId);
        const groupRef = groupRefs.get(newData.requestId);
        if (groupRef && groupRef.current) {
            await groupRef.current.addData(newData, needPurge);
        }
    }

    /**
     * Moves existing data for the old request ID to a new ID and sends the underlying group a call to
     * note that next incoming data has to replace existing data.
     * We don't remove existing data here, to avoid flickering.
     *
     * @param oldRequestId The ID under which an existing group is registered.
     * @param newRequestId The ID under which this group is accessible after return.
     */
    public reassignData(oldRequestId: string, newRequestId: string): void {
        const { groupRefs, purgePending } = this.state;

        const groupRef = groupRefs.get(oldRequestId);
        if (groupRef && groupRef.current) {
            groupRefs.delete(oldRequestId);
            groupRefs.set(newRequestId, groupRef);

            purgePending.add(newRequestId);
        }
    }

    public markPendingReplace(requestId: string): void {
        const { purgePending } = this.state;

        purgePending.add(requestId);
    }

    private handleTabSelection = (id: string): void => {
        this.setState({ selectedTab: id });
    };

}
