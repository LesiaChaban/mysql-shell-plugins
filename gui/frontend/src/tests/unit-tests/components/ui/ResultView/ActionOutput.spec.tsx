/*
 * Copyright (c) 2021, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { mount } from "enzyme";
import React from "react";
import { ActionOutput } from "../../../../../components/ResultView/ActionOutput";
import { MessageType } from "../../../../../app-logic/Types";
import { ResultStatus } from "../../../../../components/ResultView";
import { render } from "@testing-library/preact";

describe("Action output tests", (): void => {
    it("Action output elements", () => {
        const component = mount(
            <ActionOutput
                id="actionOutput1"
                className="actionOutput"
                text="Lorem ipsum dolor sit amet"
            />,
        );

        expect(component).toBeTruthy();
        expect(component.find("label")).toHaveLength(1);
        expect(component.find(ResultStatus)).toHaveLength(0);
    });

    it("Action output (Snapshot) 1", () => {
        const component = render(
            <ActionOutput
                id="actionOutput1"
                className="actionOutput"
                text="Lorem ipsum dolor sit amet"
            />,
        );
        expect(component).toMatchSnapshot();
    });

    it("Action output (Snapshot) 2", () => {
        const component = render(
            <ActionOutput
                id="actionOutput1"
                className="actionOutput"
                text="Lorem ipsum `dolor` sit amet"
                language={"markdown"}
            />,
        );
        expect(component).toMatchSnapshot();
    });

    it("Action output (Snapshot) 3", () => {
        const component = render(
            <ActionOutput
                id="actionOutput1"
                className="actionOutput"
                text="Lorem ipsum dolor sit amet"
                executionInfo={{
                    type: MessageType.Info,
                    text: "Info Message",
                }}
            />,
        );
        expect(component).toMatchSnapshot();
    });

});
