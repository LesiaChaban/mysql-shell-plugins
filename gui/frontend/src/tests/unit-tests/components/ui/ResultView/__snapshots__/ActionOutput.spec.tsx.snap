// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action output tests Action output (Snapshot) 1 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    theme="dark"
  >
    <div>
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg container fixedScrollbar resultText actionOutput"
        style="flex-direction: column;"
      >
        <label
          __self="[object Object]"
          __source="[object Object]"
          class="msg label"
          data-tooltip=""
        >
          Lorem ipsum dolor sit amet
        </label>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="msg container fixedScrollbar resultText actionOutput"
      style="flex-direction: column;"
    >
      <label
        __self="[object Object]"
        __source="[object Object]"
        class="msg label"
        data-tooltip=""
      >
        Lorem ipsum dolor sit amet
      </label>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Action output tests Action output (Snapshot) 2 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    theme="dark"
  >
    <div>
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg container fixedScrollbar resultText actionOutput"
        style="flex-direction: column;"
      >
        <label
          __self="[object Object]"
          __source="[object Object]"
          class="msg label Auto"
          data-lang="markdown"
          data-tooltip=""
        >
          Lorem ipsum \`dolor\` sit amet
        </label>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="msg container fixedScrollbar resultText actionOutput"
      style="flex-direction: column;"
    >
      <label
        __self="[object Object]"
        __source="[object Object]"
        class="msg label Auto"
        data-lang="markdown"
        data-tooltip=""
      >
        Lorem ipsum \`dolor\` sit amet
      </label>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Action output tests Action output (Snapshot) 3 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    theme="dark"
  >
    <div>
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg container fixedScrollbar resultText actionOutput"
        style="flex-direction: column;"
      >
        <label
          __self="[object Object]"
          __source="[object Object]"
          class="msg label"
          data-tooltip=""
        >
          Lorem ipsum dolor sit amet
        </label>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar resultStatus containsMessage"
          style="flex-direction: row;"
        >
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg message info"
          >
            Info Message
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="msg container fixedScrollbar resultText actionOutput"
      style="flex-direction: column;"
    >
      <label
        __self="[object Object]"
        __source="[object Object]"
        class="msg label"
        data-tooltip=""
      >
        Lorem ipsum dolor sit amet
      </label>
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg container fixedScrollbar resultStatus containsMessage"
        style="flex-direction: row;"
      >
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg message info"
        >
          Info Message
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
